experiments/2018-05-23-rust-on-android-JNA.md

```diff
@@ -0,0 +1,198 @@
+
+![Create new interface](assets/2018-05-23-rust-on-android-JNA/jna-interface.png)
+
+Add the following code to the inteface. Your interface needs to extend `com.sun.jna.Library`, which is the JNA base interface that derives all native library definitions from the instance of your library. 
+
+```
+package com.mozilla.greetings;
+
+import com.sun.jna.Library;
+import com.sun.jna.Native;
+import com.sun.jna.NativeLibrary;
+
+public interface JNA extends Library {
+    String JNA_LIBRARY_NAME = "greetings";
+    NativeLibrary JNA_NATIVE_LIB = NativeLibrary.getInstance(JNA_LIBRARY_NAME);
+
+    JNA INSTANCE = (JNA) Native.loadLibrary(JNA_LIBRARY_NAME, JNA.class);
+}
+```
+
+Here we are declaring that our library is called `greetings`. You will remember that our library is actually called `libgreetings.so`. When Java loads native libraries it will automatically prepend `lib` and append `.so` onto whatevery you call the library during your import. If you were to enter `libgreetings.so` here, the library loader will look for a file called `liblibgreetings.so.so` so be careful that you don't rename your library to something else.
+
+In is inside this interface that we will declare the headers that will link to our Rust library. We have only one external function to declare here, the `rust_greeting` function. Add the following interface declaration to your `JNA` interface.
+
+```
+String rust_greeting(String pattern);
+```
+
+That's it. We can now create some code that will call it.
+
+
+As with iOS, we're going to create a wrapper class to wrap the JNA binding and ensure that we have a more ergonmic Java API. In the project explorer on the left hand side of the studio window, ensure that `app > java > <domain>.greetings` is highlighted then go to `File > New > Java Class`. Name your class `RustGreetings` and click `OK`.
+
+![Create new class](assets/2017-09-21-rust-on-android/new-class.png)
+
+In your new class file, add the following code. Here we are defining the native interface to our Rust library and calling it `greeting`, with the same signature. The `sayHello` method simply makes a call to that native function.
+
+```java
+public class RustGreetings {
+
+    public String sayHello(String to) {
+        return JNA.INSTANCE.rust_greeting(to);
+    }
+}
+```
+
+We need to load our Rust library when the app starts, so add the following lines below the class declaration and before the `onCreate` method in `GreetingsActivity.java`.
+
+```java
+   static {
+        System.loadLibrary("greetings");
+    }
+```
+
+This looks for a library called `greetings.so` inside the jniLibs directory and picks the correct one for the current architecture.
+
+Open `res/layout/activity-greetings.xml`. In the `Component Tree` panel, highlight the `TextField` and open the `Properties` panel. Change the `ID` in the `Properties` panel to `greetingField`. This is how we are going to refer to it from our Activity.
```

Comment:
It looks like `TextField` is a `TextView` and `Properties` is `Attributes` on my Android Studio (version 3.1.2 on macOS). Perhaps a platform-specific difference, or changes in a recent Android Studio update?

Reply:
Yup that is an update to Android Studio. Good spot.